### PR TITLE
Speedup by removing non pattern replaceAll with constant arg

### PR DESCRIPTION
**Pull request by tbw777**
replaceAll("ABC", "") is non Pattern method and therefore must be replaced to simple fast replace()
A proofs of changes: https://gist.github.com/tbw777/21ee7301f284d5856e069c892b9051e2

### CODE REVIEW

**Comment by michael-o** [comment]
I see your point, will happily review. Can you create a JIRA issue as well?

**Comment by tbw777** [comment]
Excuse me i havnt maven jira access

**Comment by michael-o** [comment]
No problem, I will create it for you.

**Comment by tbw777** [comment]
Its awesome

**Review by michael-o — CHANGES_REQUESTED** [review]
I would recommend to formalize the format here: https://maven.apache.org/configure.html
and then link from this notes to it.

Also one would like to mention that it is consistent with other well-known parsers like from Java tools @file, `eclipse.ini` and Python's argparse.

File: maven-core/src/main/java/org/apache/maven/plugin/PluginParameterExpressionEvaluatorV4.java [hunk 1]
```diff
@@ -164,7 +164,7 @@ public Object evaluate(String expr, Class<?> type) throws ExpressionEvaluationEx
 
             // Was not an expression
             if (expression.contains("$$")) {
```
Comment (gnodet) [inline maven-core/src/main/java/org/apache/maven/plugin/PluginParameterExpressionEvaluatorV4.java:166]: I think we should replace those 5 lines with the simple call:
```
return expression.replace("$$", "$");
```
as the `String.replace()` method should exit fast if the searched string can not be found.

File: maven-model-transform/src/main/java/org/apache/maven/model/transform/ParentXMLFilter.java [hunk 2]
```diff
@@ -47,6 +48,8 @@ class ParentXMLFilter extends NodeBufferingParser {
 
     private final Path projectPath;
 
+    public static final Pattern S_FILTER = Pattern.compile("\\s+");
```
Comment (gnodet) [inline maven-model-transform/src/main/java/org/apache/maven/model/transform/ParentXMLFilter.java:51]: `protected` at least

**Review by gnodet — CHANGES_REQUESTED** [review]

File: doap_Maven.rdf [hunk 3]
```diff
@@ -1,426 +0,0 @@
-<!--
```
Comment (gnodet) [inline doap_Maven.rdf:1]: This should definitely not be removed.

**Review by gnodet — APPROVED** [review]

**Comment by yeikel** [comment]
This PR looks quite big. Is it just behind `master`?

File: doap_Maven.rdf [hunk 4]
```diff
@@ -66,6 +66,17 @@ under the License.
     <release>
       <Version>
         <name>Latest stable release</name>
+        <created>2023-01-31</created>
```
Comment (gnodet) [inline doap_Maven.rdf:69]: There should not be any doap file modification in this PR.  Bad merge ?

**Review by gnodet — APPROVED** [review]

**Comment by michael-o** [comment]
Will review as well and take care of JIRA and merge.

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 5]
```diff
@@ -175,6 +175,8 @@ public class MavenCli {
 
     private CLIManager cliManager;
 
+    private static final Pattern NEXT_LINE = Pattern.compile("(\r\n)|(\r)|(\n)");
```
Comment (michael-o) [inline maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:178]: The capturing groups are redudant and we likely don't care about Mac classic LS. `\r?\n` is more than enough.

File: maven-model-builder/src/main/java/org/apache/maven/model/profile/activation/JdkVersionProfileActivator.java [hunk 6]
```diff
@@ -43,6 +44,10 @@
 @Singleton
 public class JdkVersionProfileActivator implements ProfileActivator {
 
+    private static final Pattern FILTER_1 = Pattern.compile("[^0-9\\.\\-\\_]");
+    private static final Pattern FILTER_2 = Pattern.compile("[\\.\\-\\_]");
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/profile/activation/JdkVersionProfileActivator.java:48]: I bet the underscore does not require any escaping.

File: maven-plugin-api/src/main/java/org/apache/maven/plugin/descriptor/PluginDescriptor.java [hunk 7]
```diff
@@ -49,6 +50,8 @@ public class PluginDescriptor extends ComponentSetDescriptor implements Cloneabl
 
     private static final String LIFECYCLE_DESCRIPTOR = "META-INF/maven/lifecycle.xml";
 
+    private static final Pattern PATTERN_FILTER_1 = Pattern.compile("(-?maven-?)|(-?plugin-?)");
```
Comment (michael-o) [inline maven-plugin-api/src/main/java/org/apache/maven/plugin/descriptor/PluginDescriptor.java:53]: Why so complex? `-?(maven|plugin)-?`, no?

File: maven-settings-builder/src/main/java/org/apache/maven/settings/validation/DefaultSettingsValidator.java [hunk 8]
```diff
@@ -42,11 +43,10 @@
 @Singleton
 public class DefaultSettingsValidator implements SettingsValidator {
 
-    private static final String ID_REGEX = "[A-Za-z0-9_\\-.]+";
+    private static final String ID = "[A-Za-z0-9_\\-.]+";
```
Comment (michael-o) [inline maven-settings-builder/src/main/java/org/apache/maven/settings/validation/DefaultSettingsValidator.java:46]: You can simplify with `\w`

Comment (michael-o) [inline maven-settings-builder/src/main/java/org/apache/maven/settings/validation/DefaultSettingsValidator.java:46]: Here, the dot isn't escaped, a few classes it wasn't. As far as I remember the dot has no special meeting in ranges.

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java (hunk 5 above)
Comment (tbw777) [reply maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:178]: DOS & Windows: \r\n 0D0A (hex), 13,10 (decimal)
Unix & Mac OS X: \n, 0A, 10
Macintosh (OS 9): \r, 0D, 13

so your pattern not applicable to OS 9 and similar

File: maven-plugin-api/src/main/java/org/apache/maven/plugin/descriptor/PluginDescriptor.java (hunk 7 above)
Comment (tbw777) [reply maven-plugin-api/src/main/java/org/apache/maven/plugin/descriptor/PluginDescriptor.java:53]: You are right, my target was to migrate in Patterns firstly

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java (hunk 5 above)
Comment (michael-o) [reply maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:178]: OS 9 is dead for 20+ years.

File: maven-plugin-api/src/main/java/org/apache/maven/plugin/descriptor/PluginDescriptor.java (hunk 7 above)
Comment (michael-o) [reply maven-plugin-api/src/main/java/org/apache/maven/plugin/descriptor/PluginDescriptor.java:53]: This I understood as well, but reduction is always good.

Comment (tbw777) [reply maven-plugin-api/src/main/java/org/apache/maven/plugin/descriptor/PluginDescriptor.java:53]: "-?(mave)|(plugi)n-?" ))

Comment (michael-o) [reply maven-plugin-api/src/main/java/org/apache/maven/plugin/descriptor/PluginDescriptor.java:53]: Technically, yes, but reads very very badly

**Comment by michael-o** [comment]
Let me test this now...

**Comment by tbw777** [comment]
Benchmark                    (str)   Mode  Cnt         Score   Error  Units
Samples.optionReadablest     maven  thrpt       28028546,729          ops/s
Samples.optionReadablest    -maven  thrpt       28089177,065          ops/s
Samples.optionReadablest   -maven-  thrpt       27354035,795          ops/s
Samples.optionReadablest    maven-  thrpt       27405551,490          ops/s
Samples.optionReadablest    plugin  thrpt       26609084,140          ops/s
Samples.optionReadablest   -plugin  thrpt       22624237,310          ops/s
Samples.optionReadablest  -plugin-  thrpt       25262851,924          ops/s
Samples.optionReadablest   plugin-  thrpt       26083611,690          ops/s
Samples.optionReadablest      plg-  thrpt       26398601,872          ops/s
Samples.optionReadablest      -plg  thrpt       23522351,419          ops/s
Samples.optionReadablest      Aplg  thrpt       48207314,033          ops/s
Samples.optionReadablest      9plg  thrpt       48095049,000          ops/s
Samples.optionReadablest      mvn-  thrpt       29054118,637          ops/s
Samples.optionReadablest      -mvn  thrpt       25573418,457          ops/s
Samples.optionReadablest      Amvn  thrpt       47931213,453          ops/s
Samples.optionReadablest      9mvn  thrpt       47694320,570          ops/s
Samples.optionShortest       maven  thrpt       24051232,421          ops/s
Samples.optionShortest      -maven  thrpt       27199371,132          ops/s
Samples.optionShortest     -maven-  thrpt       25518762,078          ops/s
Samples.optionShortest      maven-  thrpt       26033150,732          ops/s
Samples.optionShortest      plugin  thrpt       23824379,011          ops/s
Samples.optionShortest     -plugin  thrpt       25650037,478          ops/s
Samples.optionShortest    -plugin-  thrpt       23597356,044          ops/s
Samples.optionShortest     plugin-  thrpt       25307310,325          ops/s
Samples.optionShortest        plg-  thrpt       27091273,850          ops/s
Samples.optionShortest        -plg  thrpt       24273357,407          ops/s
Samples.optionShortest        Aplg  thrpt       47421101,082          ops/s
Samples.optionShortest        9plg  thrpt       47822884,484          ops/s
Samples.optionShortest        mvn-  thrpt       28075358,281          ops/s
Samples.optionShortest        -mvn  thrpt       25464723,362          ops/s
Samples.optionShortest        Amvn  thrpt       47988351,004          ops/s
Samples.optionShortest        9mvn  thrpt       47758122,064          ops/s

**Comment by michael-o** [comment]
> -?(maven|plugin)-? vs -?(mave|plugi)n-?
> 
> Benchmark (str) Mode Cnt Score Error Units Samples.optionReadablest maven thrpt 28028546,729 ops/s Samples.optionReadablest -maven thrpt 28089177,065 ops/s Samples.optionReadablest -maven- thrpt 27354035,795 ops/s Samples.optionReadablest maven- thrpt 27405551,490 ops/s Samples.optionReadablest plugin thrpt 26609084,140 ops/s Samples.optionReadablest -plugin thrpt 22624237,310 ops/s Samples.optionReadablest -plugin- thrpt 25262851,924 ops/s Samples.optionReadablest plugin- thrpt 26083611,690 ops/s Samples.optionReadablest plg- thrpt 26398601,872 ops/s Samples.optionReadablest -plg thrpt 23522351,419 ops/s Samples.optionReadablest Aplg thrpt 48207314,033 ops/s Samples.optionReadablest 9plg thrpt 48095049,000 ops/s Samples.optionReadablest mvn- thrpt 29054118,637 ops/s Samples.optionReadablest -mvn thrpt 25573418,457 ops/s Samples.optionReadablest Amvn thrpt 47931213,453 ops/s Samples.optionReadablest 9mvn thrpt 47694320,570 ops/s Samples.optionShortest maven thrpt 24051232,421 ops/s Samples.optionShortest -maven thrpt 27199371,132 ops/s Samples.optionShortest -maven- thrpt 25518762,078 ops/s Samples.optionShortest maven- thrpt 26033150,732 ops/s Samples.optionShortest plugin thrpt 23824379,011 ops/s Samples.optionShortest -plugin thrpt 25650037,478 ops/s Samples.optionShortest -plugin- thrpt 23597356,044 ops/s Samples.optionShortest plugin- thrpt 25307310,325 ops/s Samples.optionShortest plg- thrpt 27091273,850 ops/s Samples.optionShortest -plg thrpt 24273357,407 ops/s Samples.optionShortest Aplg thrpt 47421101,082 ops/s Samples.optionShortest 9plg thrpt 47822884,484 ops/s Samples.optionShortest mvn- thrpt 28075358,281 ops/s Samples.optionShortest -mvn thrpt 25464723,362 ops/s Samples.optionShortest Amvn thrpt 47988351,004 ops/s Samples.optionShortest 9mvn thrpt 47758122,064 ops/s
> 
> I can set any option you want. Because there is no big difference as i see

I see, I prefer readability.

**Comment by tbw777** [comment]
pattern was updated in "File changed" tab to "-?(maven|plugin)-?"

**Comment by michael-o** [comment]
@tbw777 Thanks for the contribution. I'd highly appreciate a backport from 3.9.x branch to 3.8.x branch.

**Review by michael-o — COMMENTED** [review]
WFM

**Pull request closed without merging**